repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala

```diff
@@ -299,6 +304,21 @@ class KeystoneV2Filter @Inject()(configurationService: ConfigurationService,
       }
     }
 
+    def doUserTokensUpdate(userId: String, authToken: String, ttl: Int): Unit = {
+      if (CacheInvalidationFeedListener.invalidationEnabled()) {
+        //Have to use Vector, because List isn't serializeable in 2.10
+        val oldTokens = Option(datastore.get(s"$USER_ID_KEY_PREFIX$userId").asInstanceOf[Vector[String]]).getOrElse(Vector.empty[String])
+        if (!oldTokens.contains(authToken)) {
+          // @TODO: Do we need to clean up the oldTokens which may have irrelevant values.
+          // @TODO: Nothing is wrong with the data, but it may have extra values that would try to be removed later.
+          val newTokens = oldTokens ++ Vector(authToken)
+          // Updated for later cache invalidation via Atom Feed User events.
+          datastore.remove(s"$USER_ID_KEY_PREFIX$userId")
+          datastore.put(s"$USER_ID_KEY_PREFIX$userId", newTokens, ttl, TimeUnit.SECONDS)
```

Comment:
i was going to say just do a put, but there's a bigger overall issue, if we get multiple requests concurrently for the same user but different tokens they will potentially stomp each other and info will get lost. I think the right choice is to do a patch here, then you have no need to do the get either. Make it a set of some sort instead of a vector and just blindly fire off a patch.